### ISSUE TITLE
Bug: [no-invalid-void-type] Conflict with rule unbound-method

### ISSUE THREAD

**Issue by sviat9440** [typescript-eslint/typescript-eslint#6029]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=MYGwhgzhAECC0G8CwAoa0AOBXARiAlsNALYCmALgBYD2AJgBQCUi0AvquyqsNQHYTkSFGrQBc0JtAC8APmgA3avlrTovUgHc4TAHRkqdANyogA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6KImjQO0SABpw2PJPJUa9ZIxatYTAEYdDwgLaJ8ACw6DpkWfMWqAviFdA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUjoAHoizJ46IngCGqACL9YmNXl3Cwo9FIC+IM0A

### Repro Code

```typescript
class A {
  public method() { }
}

const method: () => void = new A().method;
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-invalid-void-type": "error",
    "@typescript-eslint/unbound-method": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true,
    "experimentalDecorators": true
  }
}
```

### Expected Result

```
class A {
  public method(this: void) { }
}

const method: () => void = new A().method;
```

This code should pass linting.

### Actual Result

This code:
```
class A {
  public method() { }
}

const method: () => void = new A().method;
```
causes an error: 
```
@typescript-eslint/unbound-method:
Avoid referencing unbound methods which may cause unintentional scoping of `this`.
If your function does not access `this`, you can annotate it with `this: void`, or consider using an arrow function instead. 5:28 - 5:42`
```

So, It follows from this message that if you add the annotation `this: void`, then everything should work?
I tried to change the code to this:
```
class A {
  public method(this: void) { }
}

const method: () => void = new A().method;
```
But this code causes another error:
```
@typescript-eslint/no-invalid-void-type:
void is only valid as a return type or generic type argument.
```

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.43.0` |
| `@typescript-eslint/parser` | `5.43.0` |
| `TypeScript` | `4.8.4` |
| `ESLint` | `8.15.0` |
| `node` | `web` |
